compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2006,12 +2080,67 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     err.emit()
                 } else if let Err(reported) = qself_ty.error_reported() {
                     reported
+                } else if let ty::Alias(ty::Opaque, alias_ty) = qself_ty.kind() {
+                    // `<impl Trait as OtherTrait>::Assoc` makes no sense.
+                    struct_span_err!(
+                        tcx.sess,
+                        tcx.def_span(alias_ty.def_id),
+                        E0667,
+                        "`impl Trait` is not allowed in path parameters"
+                    )
+                    .emit() // Already reported in an earlier stage.
                 } else {
+                    // Find all the `impl`s that `qself_ty` has for any trait that has the
+                    // associated type, so that we suggest the right one.
+                    let infcx = tcx.infer_ctxt().build();
+                    // We create a fresh `ty::ParamEnv` instead of the one for `self.item_def_id()`
+                    // to avoid a cycle error in `src/test/ui/resolve/issue-102946.rs`.
+                    let param_env = ty::ParamEnv::new(
+                        ty::List::empty(),
+                        traits::Reveal::All,
+                        hir::Constness::NotConst,
+                    );
+                    let traits: Vec<_> = self
+                        .tcx()
+                        .all_traits()
+                        .filter(|trait_def_id| {
+                            // Consider only traits with the associated type
+                            tcx.associated_items(*trait_def_id)
+                                .in_definition_order()
+                                .find(|i| {
+                                    i.kind.namespace() == Namespace::TypeNS
+                                        && i.ident(tcx).normalize_to_macros_2_0() == assoc_ident
+                                        && matches!(i.kind, ty::AssocKind::Type)
+                                })
+                                .is_some()
+                            // Consider only accessible traits
+                            && tcx.visibility(*trait_def_id)
+                                .is_accessible_from(self.item_def_id(), tcx)
+                            && tcx.all_impls(*trait_def_id)
+                                .filter(|impl_def_id| {
+                                    let trait_ref = tcx.impl_trait_ref(impl_def_id);
```

Comment:
This will not succeed for impls with any generic parameters, is that what you want?

Reply:
If this *is* what you want, then you could probably just filter under `has_non_region_param`, erase regions, and use `Ty == Ty` instead of having to make an inference context.

Reply:
What do you mean it won't work?

```rust
trait Trait<T> {
    type Assoc;
}
struct S;
impl<T> Trait<T> for S {
    type Assoc = ();
}

fn main() {
    let _: Trait::Assoc;
}
```
```
gh-estebank@dev-desktop-us-2:~/rust$ rustc +dev f5.rs
error[E0223]: ambiguous associated type
  --> f5.rs:10:12
   |
10 |     let _: Trait::Assoc;
   |            ^^^^^^^^^^^^ help: use the fully-qualified path: `<S as Trait>::Assoc`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0223`.
```
```rust
trait Trait<T> {
    type Assoc;
}
struct S;
impl<T> Trait<T> for S {
    type Assoc = ();
}

fn main() {
    let _: <S as Trait>::Assoc;
}
```
```
gh-estebank@dev-desktop-us-2:~/rust$ rustc +dev f5.rs
error[E0107]: missing generics for trait `Trait`
  --> f5.rs:10:18
   |
10 |     let _: <S as Trait>::Assoc;
   |                  ^^^^^ expected 1 generic argument
   |
note: trait defined here, with 1 generic parameter: `T`
  --> f5.rs:1:7
   |
1  | trait Trait<T> {
   |       ^^^^^ -
help: add missing generic argument
   |
10 |     let _: <S as Trait<T>>::Assoc;
   |                       +++

error: aborting due to previous error
```

Reply:
We could argue that the suggestion should include the type param, but it is late and the subsequent suggestion takes you in to the right place.

Reply:
That example you gave doesn't actually hit this branch -- it hits the other logic you added below starting with 2255. This is only the case when we have a qself type.

Reply:
What I meant though is that this code specifically fails for the case of an impl self ty with generic parameters in it:

```
trait Trait<A> {}

trait Assoc {
    type Ty;
}

impl<A> Assoc for dyn Trait<A> {
    type Assoc = i32;
}

fn main() {
    let x: <dyn Trait<i32>>::Ty;
}
```

Which I believe should suggest "use the fully-qualified path: `<dyn Trait<i32> as Assoc>::Ty`", but instead suggests: 

```
help: if there were a trait named `Example` with associated type `Ty` implemented for `dyn Trait<i32>`, you could use the fully-qualified path
   |
12 |     let x: <dyn Trait<i32> as Example>::Ty;
   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

You can make this happen by using `tcx.bound_impl_trait_ref` and `infcx.fresh_substs_for_item(impl_def_id)` instead, so we put inference vars into the self type correctly.

Reply:
done!